### ISSUE TITLE
iszero

### ISSUE THREAD

**Issue by blegat** [JuliaAlgebra/StarAlgebras.jl#60]
Not high priority but this won't work if the `SparseCoefficients` is not canonicalized:

https://github.com/JuliaAlgebra/StarAlgebras.jl/blob/ba3697f45b957b1881950aa096932a42b1cc7c9b/src/coefficients.jl#L37

This doesn't work for tuples:

https://github.com/JuliaAlgebra/StarAlgebras.jl/blob/ba3697f45b957b1881950aa096932a42b1cc7c9b/src/types.jl#L92

**Comment by kalmarek** [comment]
Yes, we need to fix our semantics and develop a consistent mindset for dealing with canonicalization.
1. Is is a temporary state within `unsafe_` functions?
2. If yes, do we bring the coefficients to canonical form upon construction, or should we trust the user?
3. does canonical always perform its task, or is is a no-op (and hence we can call it freely in our code) for already canonicalized coeffs?

especially the latter will experience problems with thread-safety, but maybe its not our goal?

**Comment by blegat** [comment]
You can only bring it to a non-canonical state if we call `unsafe_` function of `UnsafeAddMul` then you are responsible to call `canonicalize!`. I don't think it's a no-op then, we need `canonicalize!` to do its thing right. If the user calls it, it means the user know that he did unsafe things so he's calling it explicitly. If the user never calls unsafe methods then he should never have to canonicalize. If he does, then it's not a no-op and there is a cost, so he should just not do it ^^
